File: SBSAnimoji/Pages/Main/SBSPuppetView.m
//
//  SBSPuppetView.m
//  SBSAnimoji
//
//  Created by Simon Støvring on 06/11/2017.
//  Copyright © 2017 SimonBS. All rights reserved.
//

#import "SBSPuppetView.h"
#import <objc/runtime.h>

#define MAX_RECORDING_DURATION 60 // Seconds

@implementation SBSPuppetView

- (void)audioPlayerItemDidReachEnd:(id)arg1 {
    [super audioPlayerItemDidReachEnd:arg1];
    if ([self.sbsDelegate respondsToSelector:@selector(puppetViewDidFinishPlaying:)]) {
        [self.sbsDelegate puppetViewDidFinishPlaying:self];
    }
}

- (void)startRecording {
    [super startRecording];
    
    int duration = MAX_RECORDING_DURATION * 60;
    
    NSMutableData *timesBuffer = [NSMutableData dataWithCapacity: duration * 8];
    NSMutableData *blendShapeBuffer = [NSMutableData dataWithCapacity: duration * 204];
    NSMutableData *transformData = [NSMutableData dataWithCapacity: duration * 64];
    
    [self setValue:[NSNumber numberWithInt:duration] forKey:@"_recordingCapacity"];
    [self setValue:timesBuffer forKey:@"_rawTimesData"];
    [self setValue:blendShapeBuffer forKey:@"_rawBlendShapesData"];
    [self setValue:transformData forKey:@"_rawTransformsData"];
    
    {
        Ivar ivar = class_getInstanceVariable([AVTRecordView class], "_rawTimes");
        object_setIvar(self, ivar, [timesBuffer mutableBytes]);
    }
    
    {
        Ivar ivar = class_getInstanceVariable([AVTRecordView class], "_rawBlendShapes");
        object_setIvar(self, ivar, [blendShapeBuffer mutableBytes]);
    }
    
    {
        Ivar ivar = class_getInstanceVariable([AVTRecordView class], "_rawTransforms");
        object_setIvar(self, ivar, [transformData mutableBytes]);
    }
    
    if ([self.sbsDelegate respondsToSelector:@selector(puppetViewDidStartRecording:)]) {
        [self.sbsDelegate puppetViewDidStartRecording:self];
    }
}

- (void)stopRecording {
    [super stopRecording];
    if ([self.sbsDelegate respondsToSelector:@selector(puppetViewDidStopRecording:)]) {
        [self.sbsDelegate puppetViewDidStopRecording:self];
    }
}

@end

File: Animoji/Model/SBSPuppetView.swift
//
//  SBSPuppetView.swift
//  Animoji
//
//  Created by Daniel Illescas Romero on 25/12/2017.
//  Copyright © 2017 Daniel Illescas Romero. All rights reserved.
//

import UIKit

@objc protocol SBSPuppetViewDelegate: NSObjectProtocol {
	func puppetViewDidFinishPlaying(puppetView: SBSPuppetView)
	func puppetViewDidStartRecording(puppetView: SBSPuppetView)
	func puppetViewDidStopRecording(puppetView: SBSPuppetView)
}

@objc class SBSPuppetView: AVTPuppetView {
	
	var sbsDelegate: SBSPuppetViewDelegate?
	
	override public func audioPlayerItemDidReachEnd(_ arg1: Any!) {
		super.audioPlayerItemDidReachEnd(arg1)
		if self.sbsDelegate?.responds(to: #selector(self.sbsDelegate?.puppetViewDidFinishPlaying(puppetView:))) == true {
			self.sbsDelegate?.puppetViewDidFinishPlaying(puppetView: self)
		}
	}
	
	override func startRecording() {
		super.startRecording()
		
		let recordingDuration = 60 // seconds
		
		let duration = recordingDuration * 60
		
		let timesBuffer = NSMutableData(capacity: duration * 8)
		let blendShapeBuffer = NSMutableData(capacity: duration * 204)
		let transformData = NSMutableData(capacity: duration * 64)
		
		self.setValue(duration, forKey: "_recordingCapacity")
		self.setValue(timesBuffer, forKey: "_rawTimesData")
		self.setValue(blendShapeBuffer, forKey: "_rawBlendShapesData")
		self.setValue(transformData, forKey: "_rawTransformsData")
		
		if let ivarRawTimes = class_getInstanceVariable(AVTPuppetView.self, "_rawTimes") {
			object_setIvar(self, ivarRawTimes, timesBuffer?.mutableBytes)
		}
		
		if let ivarBlendShapes = class_getInstanceVariable(AVTPuppetView.self, "_rawBlendShapes") {
			object_setIvar(self, ivarBlendShapes, blendShapeBuffer?.mutableBytes)
		}
		
		if let ivarRawTransforms = class_getInstanceVariable(AVTPuppetView.self, "_rawTransforms") {
			object_setIvar(self, ivarRawTransforms, transformData?.mutableBytes)
		}
		
		if self.sbsDelegate?.responds(to: #selector(self.sbsDelegate?.puppetViewDidStartRecording(puppetView:))) == true {
			self.sbsDelegate?.puppetViewDidStartRecording(puppetView: self)
		}
	}
	
	override func stopRecording() {
		super.stopRecording()
		if self.sbsDelegate?.responds(to: #selector(self.sbsDelegate?.puppetViewDidStopRecording(puppetView:))) == true {
			self.sbsDelegate?.puppetViewDidStopRecording(puppetView: self)
		}
	}
}
